src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/TypeRef.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// This is consumed as 'generated' code in a source package and therefore requires an explicit nullable enable
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.CommonLanguageServerProtocol.Framework;
+
+/// <summary>
+/// Helper that avoids loading a <see cref="Type"/> by its full assembly-qualified name until needed.
+/// </summary>
+internal abstract partial class TypeRef : IEquatable<TypeRef>
+{
+    private static readonly Dictionary<string, TypeRef> s_typeNameToLazyTypeMap = [];
+
+    /// <summary>
+    /// Returns the full assembly-qualified name of this type.
+    /// </summary>
+    public string TypeName { get; }
+
+    /// <summary>
+    /// Returns the underlying <see cref="Type"/>, potentially loading its assembly.
+    /// </summary>
+    public abstract Type GetResolvedType();
+
+    private TypeRef(string typeName)
+    {
+        TypeName = typeName ?? throw new ArgumentNullException(nameof(typeName));
+    }
+
+    [return: NotNullIfNotNull(nameof(typeName))]
+    public static TypeRef? From(string? typeName)
+    {
+        if (typeName is null)
+        {
+            return null;
+        }
+
+        lock (s_typeNameToLazyTypeMap)
+        {
+            if (!s_typeNameToLazyTypeMap.TryGetValue(typeName, out var result))
+            {
+                result = new LazyTypeRef(typeName);
+                s_typeNameToLazyTypeMap.Add(typeName, result);
+            }
+
+            return result;
+        }
+    }
+
+    [return: NotNullIfNotNull(nameof(type))]
+    public static TypeRef? From(Type? type)
+    {
+        if (type is null)
+        {
+            return null;
+        }
+
+        lock (s_typeNameToLazyTypeMap)
+        {
+            var typeName = type.AssemblyQualifiedName!;
```

Comment:
Wondering if we might have an issue here when the same type is  loaded into different ALCs.  This is a use case for the VSCode server, where we load third party LSP extensions into a separate assembly load context (currently its only Fred's compiler SDK extension as we haven't made the APIs fully public).

If the same assembly is loaded twice (once into each ALC), what happens to the `s_typeNameToLazyTypeMap` (are the `AssemblyQualifiedName` different and hence two instances in the map), and when resolving the actual type - does it resolve to the type in the correct assembly load context?  Or does it (incorrectly?) pick the type from the default context always.

Reply:
If a TypeRef for the same Type is created from different ALC, I would expect the following things to be true:

1. The `AssemblyQualifiedName` would be the same across ALCs.
2. There would be multiple instances of `s_typeNameToLazyTypeMap`, one for each ALC, since statics aren't shared across ALCs. So, there would be two `TypeRefs` for the same Type across different ALCs, and I think that's definitely helpful for this scenario.

All of that said, I agree that there's a problem and a change is needed. The problem is `Type.GetType(...)`, since it can load assemblies and the ALC to load assemblies from is ambiguous. I can think of a couple solutions:

1. We can change `TypeRef` to optionally take an `IAssemblyLoader` (from VS MEF). Then, we'd ensure that the language server is constructed with a `CustomExportAssemblyLoader` in the VS Code scenario. `TypeRef` would then call the `Type.GetType(...)` overload that takes a func to resolve assemblies and use that to call into the `IAssemblyLoader` that was passed in.
2. We can wrap the call to `Type.GetType(...)` in `AssemblyLoadContext.EnterContextualReflection()` and pass the ALC into `GetResolvedType()`.

I think the first option is the better one. I don't think there's a reason to start exposing the ALCs that are created by the `ExtensionAssemblyManager`.

Do you have any thoughts on that solution? If I go implement it, I'd be grateful if somebody could help me test the VS Code scenario with extra extensions in ALCs.

Reply:
OK. I'm going to introduce an `ITypeRefResolver` interface that is used to resolve a `TypeRef` to a `Type`. For VS Code, we'll provide an `ITypeRefResolver` implementation that handles the ALC scenario.